src/lib/pubkey/frodokem/frodokem.cpp

```diff
@@ -0,0 +1,398 @@
+         // and performs a single CT comparison. For convenience we compare the
+         // matrices individually in CT and CT-&& the result.
+         //
+         // Note: b_p and c are unpacked values that are reduced by definition.
+         //       b_pp and c_p are calculated values that need the reduction for
+         //       an unambiguous comparison.
+         b_pp.reduce(consts);
+         c_p.reduce(consts);
+         const bool cmp_b_p_b_pp = b_p.constant_time_compare(b_pp);
+         const bool cmp_c_c_p = c.constant_time_compare(c_p);
+         const auto cmp = CT::Mask<uint8_t>::expand(cmp_b_p_b_pp) & CT::Mask<uint8_t>::expand(cmp_c_c_p);
+
+         std::vector<uint8_t> k_bar(consts.len_sec_bytes(), 0);
+         CT::conditional_copy_mem(
+            cmp.value(), k_bar.data(), k_p.data(), m_private_key->s().data(), consts.len_sec_bytes());
+
+         shake.update(encapsulated_key);
+         shake.update(k_bar);
+         shake.output(out_shared_key);
+      }
+
+   private:
+      std::shared_ptr<FrodoKEM_PublicKeyInternal> m_public_key;
+      std::shared_ptr<FrodoKEM_PrivateKeyInternal> m_private_key;
+};
+
+//
+// - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
+//
+
+FrodoKEM_PublicKey::FrodoKEM_PublicKey(std::span<const uint8_t> pub_key, FrodoKEMMode mode) {
+   FrodoKEMConstants consts(mode);
+   if(pub_key.size() != consts.len_public_key_bytes()) {
+      throw Invalid_Argument("FrodoKEM public key does not have the correct byte count");
+   }
+
+   BufferSlicer pk_bs(pub_key);
+   auto seed_a = pk_bs.copy<FrodoSeedA>(consts.len_a_bytes());
+   const auto packed_b = pk_bs.take<FrodoPackedMatrix>(consts.d() * consts.n() * consts.n_bar() / 8);
+   BOTAN_ASSERT_NOMSG(pk_bs.empty());
+
+   auto b = FrodoMatrix::unpack(consts, std::tuple(consts.n(), consts.n_bar()), packed_b);
+
+   m_public = std::make_shared<FrodoKEM_PublicKeyInternal>(std::move(consts), std::move(seed_a), std::move(b));
+}
+
+FrodoKEM_PublicKey::FrodoKEM_PublicKey(const AlgorithmIdentifier& alg_id, std::span<const uint8_t> key_bits) :
+      FrodoKEM_PublicKey(key_bits, FrodoKEMMode(alg_id.oid())) {}
+
+FrodoKEM_PublicKey::FrodoKEM_PublicKey(const FrodoKEM_PublicKey& other) {
+   m_public = std::make_shared<FrodoKEM_PublicKeyInternal>(
+      other.m_public->constants(), other.m_public->seed_a(), other.m_public->b());
+}
+
+FrodoKEM_PublicKey& FrodoKEM_PublicKey::operator=(const FrodoKEM_PublicKey& other) {
+   if(this != &other) {
+      m_public = std::make_shared<FrodoKEM_PublicKeyInternal>(
+         other.m_public->constants(), other.m_public->seed_a(), other.m_public->b());
+   }
+   return *this;
+}
+
+AlgorithmIdentifier FrodoKEM_PublicKey::algorithm_identifier() const {
+   return AlgorithmIdentifier(object_identifier(), AlgorithmIdentifier::USE_EMPTY_PARAM);
+}
+
+OID FrodoKEM_PublicKey::object_identifier() const {
+   return OID::from_string(m_public->constants().mode().to_string());
```

Comment:
Couldn't this use `mode().object_identifier()`?

Reply:
Fixed